onnx/defs/tensor/defs.cc

```diff
@@ -3943,4 +3943,75 @@ ONNX_OPERATOR_SET_SCHEMA(
           return true;
         }));
 
+static const char* TensorScatter_ver24_doc = R"DOC(
+TensorScatter performs kv cache updates for Attention calculations. The past and present cache tensors have the
+same shape, with the sequence length dimension being max_seqlen, so the sizes of these tensors do not need to 
+grow between iterations. 
+The optional write_indices input indicates the write index for each sample in the batch, assumed to be zero 
+if not provided. During the prefill phase of attention, only the first two inputs are needed. During the decode 
+phase, write_indices is also needed so that the incoming k and v can be appended after the last valid token 
+for each sample in the batch.
+
+In order to perform kv caching in place, as is the common practice for efficient inference, the execution 
+provider needs to alias the buffers for .
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    TensorScatter,
+    24,
+    OpSchema()
+        .SetDoc(TensorScatter_ver24_doc)
+        .Attr(
+            "mode",
+            "The write mode of kv cache. Supported modes include `linear` and `circular`. `linear` mode requires "
+            "write_indices+sequence_length<=max_sequence_length. For `circular` mode, the updates happen in "
+            "wrap-around fashion, ie, the update index is modulo `max_sequence_length`",
+            AttributeProto::STRING,
+            std::string("linear"))
+        .Input(
+            0,
+            "past_cache",
+            "Past state cache for key or value tensor with 4D shape `(batch_size, num_heads, max_sequence_length, k_head_size)`"
```

Comment:
I wonder if we should generalize this? Eg., two possibilities exist:
(1) A simpler generalization is to allow a shape (batchsize, D1, D2, ... Dn), with an axis attribute for identifying the sequence-axis, unless we want to fix sequence axis to be -2
(2) More general would allow even batchsize dimension to be multiple axes, with input of shape (bs1, bs2, ..., D1, D2, ...). It would be a bit more complicated.

Reply:
Correct, that's the complication.

Reply:
@gramalingam any thoughts?